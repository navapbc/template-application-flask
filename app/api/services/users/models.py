import dataclasses
from datetime import date

from api.db.models import user_models
from api.services.core import patch_params


@dataclasses.dataclass
class RoleParams:
    type: user_models.RoleType


@dataclasses.dataclass
class CreateUserParams:
    first_name: str | None = None
    middle_name: str | None = None
    last_name: str | None = None
    phone_number: str | None = None
    date_of_birth: date | None = None
    is_active: bool | None = None
    roles: list[RoleParams] | None = None


PatchUserParams = patch_params.PatchParams[CreateUserParams]
