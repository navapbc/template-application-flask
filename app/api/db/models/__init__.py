from . import example_person_models

__all__ = ["example_person_models"]
