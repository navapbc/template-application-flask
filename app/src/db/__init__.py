__all__ = ["bulk_ops"]
