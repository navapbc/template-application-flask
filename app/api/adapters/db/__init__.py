"""
Database adapter for Postgres database.
This module is an abstraction around the psycopg3 python postgres library. It
has methods for connecting to the database, and abstracts away the loading of
config and environment variables.

Usage:
    # During application initialization
    from api.adapters import db
    db.init()

    # In a route handler
    from api.adapters import db
    with db.get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute("SELECT 1")
"""

import atexit
import os

import psycopg
import psycopg_pool
from psycopg import rows

from api import logging

# Re-export these types to restrict the API of this module
# to only the types we want to expose explicitly.
Connection = psycopg.Connection
Cursor = psycopg.Cursor
DBConnectionError = psycopg.OperationalError
kwargs_row = rows.kwargs_row

logger = logging.get_logger(__name__)


def init():
    """Initializes the database connection pool"""
    global pool
    print(get_conninfo())
    pool = psycopg_pool.ConnectionPool(get_conninfo(), open=False)

    # Open the connection pool when the app starts
    pool.open()

    # # Close the connection pool when the app shuts down
    atexit.register(pool.close)


def get_conninfo() -> str:
    """Gets a db connection info based on environment variables"""
    dbhost = os.getenv("DB_HOST")
    dbname = os.getenv("POSTGRES_DB")
    username = os.getenv("POSTGRES_USER")
    password = os.getenv("POSTGRES_PASSWORD")
    ENVIRONMENT = os.getenv("ENVIRONMENT", "production")
    sslmode = "allow" if ENVIRONMENT == "local" else "verify-full"

    return psycopg.conninfo.make_conninfo(
        host=dbhost,
        dbname=dbname,
        user=username,
        password=password,
        sslmode=sslmode,
    )


def get_connection() -> psycopg.Connection:
    """Gets a db connection
    Uses database credentials in environment variables to obtain a database
    connection.
    Usage:
      with get_connection() as conn:
        with conn.cursor() as cur:
          cur.execute("SELECT 1")
          result = cur.fetchall()
    """
    return pool.connection()


# Global connection pool
# The simplest way to use the psycopg pool is to create a single instance of it,
# and to use this object in the rest of the program. The pool is threadsafe and
# can be used by multiple threads at the same time.
# https://www.psycopg.org/psycopg3/docs/advanced/pool.html
pool: psycopg_pool.ConnectionPool | None = None

__all__ = ["init", "get_connection", "Connection", "Cursor", "kwargs_row"]
