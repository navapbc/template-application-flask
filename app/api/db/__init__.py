import os
import urllib.parse
from contextlib import contextmanager
from typing import Any, Generator, Optional

import psycopg2
import sqlalchemy.pool as pool
from sqlalchemy import create_engine
from sqlalchemy.engine import Engine
from sqlalchemy.orm import scoped_session, sessionmaker

import api.logging
from api.db.db_config import DbConfig, get_db_config
from api.db.migrations.run import have_all_migrations_run

logger = api.logging.get_logger(__name__)


def init(
    config: Optional[DbConfig] = None, check_migrations_current: bool = False
) -> scoped_session:
    logger.info("connecting to postgres db")

    engine = create_db_engine(config)
    conn = engine.connect()

    conn_info = conn.connection.dbapi_connection.info  # type: ignore
    logger.info(
        "connected to postgres db",
        extra={
            "dbname": conn_info.dbname,
            "user": conn_info.user,
            "host": conn_info.host,
            "port": conn_info.port,
            "options": conn_info.options,
            "dsn_parameters": conn_info.dsn_parameters,
            "protocol_version": conn_info.protocol_version,
            "server_version": conn_info.server_version,
        },
    )
    verify_ssl(conn_info)

    # Explicitly commit sessions â€” usually with session_scope. Also disable expiry on commit,
    # as we don't need to be strict on consistency within our routes. Once we've retrieved data
    # from the database, we shouldn't make any extra requests to the db when grabbing existing
    # attributes.
    session_factory = scoped_session(
        sessionmaker(autocommit=False, expire_on_commit=False, bind=engine)
    )

    if check_migrations_current:
        have_all_migrations_run(engine)

    engine.dispose()

    return session_factory


def verify_ssl(connection_info: Any) -> None:
    """Verify that the database connection is encrypted and log a warning if not."""
    if connection_info.ssl_in_use:
        logger.info(
            "database connection is using SSL: %s",
            ", ".join(
                name + " " + connection_info.ssl_attribute(name)
                for name in connection_info.ssl_attribute_names
            ),
        )
    else:
        logger.warning("database connection is not using SSL")


def create_db_engine(config: Optional[DbConfig] = None) -> Engine:
    db_config: DbConfig = config if config is not None else get_db_config()

    # We want to be able to control the connection parameters for each
    # connection because for IAM authentication with RDS, short-lived tokens are
    # used as the password, and so we potentially need to generate a fresh token
    # for each connection.
    #
    # For more details on building connection pools, see the docs:
    # https://docs.sqlalchemy.org/en/13/core/pooling.html#constructing-a-pool
    def get_conn() -> Any:
        return psycopg2.connect(**get_connection_parameters(db_config))

    conn_pool = pool.QueuePool(get_conn, max_overflow=10, pool_size=20, timeout=3)

    # The URL only needs to specify the dialect, since the connection pool
    # handles the actual connections.
    #
    # (a SQLAlchemy Engine represents a Dialect+Pool)
    return create_engine(
        "postgresql://",
        pool=conn_pool,
        # FYI, execute many mode handles how SQLAlchemy handles doing a bunch of inserts/updates/deletes at once
        # https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#psycopg2-fast-execution-helpers
        executemany_mode="batch",
        hide_parameters=db_config.hide_sql_parameter_logs,
        # TODO: Don't think we need this as we aren't using JSON columns, but keeping for reference
        # json_serializer=lambda o: json.dumps(o, default=pydantic.json.pydantic_encoder),
    )


def get_connection_parameters(db_config: DbConfig) -> dict[str, Any]:
    connect_args = {}
    environment = os.getenv("ENVIRONMENT")
    if not environment:
        raise Exception("ENVIRONMENT is not set")

    if environment != "local":
        connect_args["sslmode"] = "require"

    return dict(
        host=db_config.host,
        dbname=db_config.name,
        user=db_config.username,
        password=db_config.password,
        port=db_config.port,
        options=f"-c search_path={db_config.db_schema}",
        connect_timeout=3,
        **connect_args,
    )


@contextmanager
def session_scope(
    session: scoped_session, close: bool = False
) -> Generator[scoped_session, None, None]:
    """Provide a transactional scope around a series of operations.

    See https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it
    """

    try:
        yield session
        session.commit()
    except Exception:
        session.rollback()
        raise
    finally:
        if close:
            session.close()


def make_connection_uri(config: DbConfig) -> str:
    """Construct PostgreSQL connection URI

    More details at:
    https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
    """
    host = config.host
    db_name = config.name
    username = config.username
    password = urllib.parse.quote(config.password) if config.password else None
    schema = config.db_schema
    port = config.port

    netloc_parts = []

    if username and password:
        netloc_parts.append(f"{username}:{password}@")
    elif username:
        netloc_parts.append(f"{username}@")
    elif password:
        netloc_parts.append(f":{password}@")

    netloc_parts.append(host)

    if port:
        netloc_parts.append(f":{port}")

    netloc = "".join(netloc_parts)

    uri = f"postgresql://{netloc}/{db_name}?options=-csearch_path={schema}"

    return uri
