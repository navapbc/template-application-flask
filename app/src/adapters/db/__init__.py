"""
Database module.

This module contains the DBClient class, which is used to manage database connections.
This module can be used on it's own or with an application framework such as Flask.

To use this module with Flask, use the flask_db module.

Usage:
    import src.adapters.db as db

    db_client = db.PostgresDBClient()

    # non-ORM style usage
    with db_client.get_connection() as conn:
        conn.execute(...)

    # ORM style usage
    with db_client.get_session() as session:
        session.query(...)
        with session.begin():
            session.add(...)
"""

# Re-export for convenience
from src.adapters.db.client import Connection, DBClient, Session
from src.adapters.db.clients.postgres_client import PostgresDBClient

# Do not import flask_db here, because this module is not dependent on any specific framework.
# Code can choose to use this module on its own or with the flask_db module depending on needs.

__all__ = ["Connection", "DBClient", "Session", "PostgresDBClient"]
