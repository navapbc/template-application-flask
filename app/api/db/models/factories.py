import os
import random
import unittest.mock
from datetime import datetime

import factory
import faker

import api.db as db
import api.db.models.example_person_models as example_person_models
import api.util.datetime_util as datetime_util

db_session = None

fake = faker.Faker()


def get_db_session() -> db.scoped_session:
    global db_session

    if os.getenv("DB_FACTORIES_DISABLE_DB_ACCESS", "0") == "1":
        alert_db_session = unittest.mock.MagicMock()
        alert_db_session.add.side_effect = Exception(
            """DB_FACTORIES_DISABLE_DB_ACCESS is set, refusing database action.

            If your tests don't need to cover database behavior, consider
            calling the `build()` method instead of `create()` on the factory to
            not persist the generated model.

            If running tests that actually need data in the DB, pull in the
            `initialize_factories_session` fixture.

            If running factories outside of the tests and you see this, unset
            the DB_FACTORIES_DISABLE_DB_ACCESS env var.
            """
        )
        return alert_db_session

    if db_session is None:
        db_session = db.init()

    return db_session


Session = db.scoped_session(lambda: get_db_session(), scopefunc=lambda: get_db_session())  # type: ignore


class Generators:
    Now = factory.LazyFunction(datetime.now)
    UtcNow = factory.LazyFunction(datetime_util.utcnow)
    UuidObj = factory.Faker("uuid4", cast_to=None)


class BaseFactory(factory.alchemy.SQLAlchemyModelFactory):
    class Meta:
        abstract = True
        sqlalchemy_session = Session
        sqlalchemy_session_persistence = "commit"


class ExamplePersonFactory(BaseFactory):
    class Meta:
        model = example_person_models.ExamplePerson

    example_person_id = Generators.UuidObj
    first_name = factory.Faker("first_name")
    last_name = factory.Faker("last_name")
    phone_number = "123-456-7890"
    date_of_birth = factory.Faker("date_object")
    is_real = factory.Faker("boolean")


class ExamplePetFactory(BaseFactory):
    class Meta:
        model = example_person_models.ExamplePet

    example_pet_id = Generators.UuidObj
    name = factory.Faker("first_name")
    species = random.choice(["Dog", "Cat", "Bird", "Fish"])

    pet_owner = factory.SubFactory(ExamplePersonFactory)
    pet_owner_id = factory.LazyAttribute(lambda p: p.pet_owner.example_person_id)
