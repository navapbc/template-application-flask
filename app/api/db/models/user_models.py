import enum
from datetime import date
from typing import Optional
from uuid import UUID

from sqlalchemy import Boolean, Column, Date, ForeignKey, Text
from sqlalchemy.dialects.postgresql import UUID as postgresUUID
from sqlalchemy.orm import Mapped, relationship

import api.logging

from .base import Base, TimestampMixin, uuid_gen

logger = api.logging.get_logger(__name__)


class RoleEnum(str, enum.Enum):
    USER = "User"
    ADMIN = "Admin"
    THIRD_PARTY = "Third Party"


class User(Base, TimestampMixin):
    __tablename__ = "user"

    user_id: Mapped[UUID] = Column(postgresUUID(as_uuid=True), primary_key=True, default=uuid_gen)

    first_name: str = Column(Text, nullable=False)
    middle_name: Optional[str] = Column(Text)
    last_name: str = Column(Text, nullable=False)
    phone_number: str = Column(Text, nullable=False)
    date_of_birth: date = Column(Date, nullable=False)
    is_active: bool = Column(Boolean, nullable=False)

    roles: Optional[list["UserRole"]] = relationship(
        "UserRole", uselist=True, back_populates="user"
    )


class UserRole(Base, TimestampMixin):
    __tablename__ = "user_role"
    user_id: Mapped[UUID] = Column(
        postgresUUID(as_uuid=True), ForeignKey("user.user_id"), primary_key=True
    )
    role_description: str = Column(Text, primary_key=True, nullable=False, index=True)

    user: User = relationship(User, back_populates="roles")
