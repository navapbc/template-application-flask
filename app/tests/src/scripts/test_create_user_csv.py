import os
import os.path as path
import re

import flask.testing
import pytest
from pytest_lazyfixture import lazy_fixture
from smart_open import open as smart_open

import src.adapters.db as db
from src.db.models.user_models import User
from tests.src.db.models.factories import UserFactory


@pytest.fixture
def prepopulate_user_table(enable_factory_create, db_session: db.Session) -> list[User]:
    # First make sure the table is empty, as other tests may have inserted data
    # and this test expects a clean slate (unlike most tests that are designed to
    # be isolated from other tests)
    db_session.query(User).delete()
    return [
        UserFactory.create(first_name="Jon", last_name="Doe", is_active=True),
        UserFactory.create(first_name="Jane", last_name="Doe", is_active=False),
        UserFactory.create(
            first_name="Alby",
            last_name="Testin",
            is_active=True,
        ),
    ]


@pytest.fixture
def tmp_s3_folder(mock_s3_bucket):
    return f"s3://{mock_s3_bucket}/path/to/"


@pytest.mark.parametrize(
    "dir",
    [
        pytest.param(lazy_fixture("tmp_s3_folder"), id="write-to-s3"),
        pytest.param(lazy_fixture("tmp_path"), id="write-to-local"),
    ],
)
def test_create_user_csv(
    prepopulate_user_table: list[User],
    cli_runner: flask.testing.FlaskCliRunner,
    dir: str,
):
    cli_runner.invoke(args=["user", "create-csv", "--dir", dir, "--filename", "test.csv"])
    output = smart_open(path.join(dir, "test.csv")).read()
    expected_output = open(
        path.join(path.dirname(__file__), "test_create_user_csv_expected.csv")
    ).read()
    assert output == expected_output


def test_default_filename(cli_runner: flask.testing.FlaskCliRunner, tmp_path: str):
    cli_runner.invoke(args=["user", "create-csv", "--dir", tmp_path])
    filenames = os.listdir(tmp_path)
    assert re.match(r"\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}-user-roles.csv", filenames[0])
