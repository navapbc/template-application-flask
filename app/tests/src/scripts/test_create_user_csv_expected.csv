"User Name","Roles","Is User Active?"
"Jon Doe","ADMIN USER","True"
"Jane Doe","ADMIN USER","False"
"Alby Testin","ADMIN USER","True"
